cerberus-web/src/main/java/com/nike/cerberus/service/JwtService.java

```diff
@@ -0,0 +1,192 @@
+    Jws<Claims> claimsJws;
+    try {
+      claimsJws =
+          Jwts.parser()
+              .requireIssuer(environmentName)
+              .setSigningKeyResolver(signingKeyResolver)
+              .parseClaimsJws(token);
+    } catch (InvalidClaimException e) {
+      log.warn("Invalid claim when parsing token: {}", token, e);
+      return Optional.empty();
+    } catch (JwtException e) {
+      log.warn("Error parsing JWT token: {}", token, e);
+      return Optional.empty();
+    } catch (IllegalArgumentException e) {
+      log.warn("Error parsing JWT token: {}", token, e);
+      return Optional.empty();
+    }
+    Claims claims = claimsJws.getBody();
+    if (blocklist.contains(claims.getId())) {
+      log.warn("This JWT token is blocklisted. ID: {}", claims.getId());
+      return Optional.empty();
+    }
+    String subject = claims.getSubject();
+    CerberusJwtClaims cerberusJwtClaims =
+        new CerberusJwtClaims()
+            .setId(claims.getId())
+            .setPrincipal(subject)
+            .setExpiresTs(
+                OffsetDateTime.ofInstant(
+                    claims.getExpiration().toInstant(), ZoneId.systemDefault()))
+            .setCreatedTs(
+                OffsetDateTime.ofInstant(claims.getIssuedAt().toInstant(), ZoneId.systemDefault()))
+            .setPrincipalType(claims.get(PRINCIPAL_TYPE_CLAIM_NAME, String.class))
+            .setGroups(claims.get(GROUP_CLAIM_NAME, String.class))
+            .setIsAdmin(claims.get(IS_ADMIN_CLAIM_NAME, Boolean.class))
+            .setRefreshCount(claims.get(REFRESH_COUNT_CLAIM_NAME, Integer.class));
+
+    return Optional.of(cerberusJwtClaims);
+  }
+
+  /** Refresh signing keys in {@link CerberusSigningKeyResolver} */
+  public void refreshKeys() {
+    signingKeyResolver.refresh();
+  }
+
+  /** Refresh JWT blocklist */
+  public void refreshBlocklist() {
+    blocklist = jwtBlocklistDao.getBlocklist();
+  }
+
+  /**
+   * Revoke JWT
+   *
+   * @param id JWT ID
+   * @param tokenExpires Expiration timestamp of the JWT
+   */
+  public void revokeToken(String id, OffsetDateTime tokenExpires) {
+    blocklist.add(id);
+    JwtBlocklistRecord jwtBlocklistRecord =
+        new JwtBlocklistRecord().setId(id).setExpiresTs(tokenExpires);
+    jwtBlocklistDao.addToBlocklist(jwtBlocklistRecord);
+  }
+
+  /**
+   * Delete JWT blocklist entries that have expired
+   *
+   * @return
+   */
+  @Transactional(
+      isolation = READ_UNCOMMITTED // allow dirty reads so we don't block other threads
+      )
+  public int deleteExpiredTokens() {
+    return jwtBlocklistDao.deleteExpiredTokens();
+  }
+
+  /**
+   * Return if the token looks like a JWT. Technically a JWT can have one dot but we don't allow it
+   * here.
+   *
+   * @param token The token to examine
+   * @return Does the token look like a JWT
+   */
+  public boolean isJwt(String token) {
+    return StringUtils.countMatches(token, SEPARATOR_CHAR) == 2;
```

Comment:
What does 2 represent in this case?

Reply:
A JWT string is 3 dot separated parts.

Reply:
There are three parts separated by '.' characters if token is a JWT token.

Reply:
e.g. eyJhbGciOiJIUzI1NiJ9.eyJzdWIiOiJKb2UifQ.ipevRNuRP6HflG8cFKnmUPtypruRC4fb1DWtoLL62SY from https://www.jsonwebtoken.io/

Reply:
I'll make a constant for it.